src/lib/data/alignment-group.js

```diff
@@ -122,36 +157,179 @@ export default class AlignmentGroup {
+      if (tokenIndex >= 0) {
+        this.origin.splice(tokenIndex, 1)
+      }
+    }
+
+    for (let i = 0; i < tokensGroup.target.length; i++) {
+      const tokenIdWord = tokensGroup.target[i]
+      const tokenIndex = this.target.findIndex(tokenId => tokenId === tokenIdWord)
+      if (tokenIndex >= 0) {
+        this.target.splice(tokenIndex, 1)
+      }
+    }
+
+    this.defineFirstStepToken()
+
+    return {
+      tokensGroup,
+      indexDeleted: step.indexDeleted
+    }
+  }
+
+  /**
+   * Finds token with the given idWord using steps
+   * @param {String} idWord
+   * @retuns { Token|null }
+   */
+  findTokenByIdWord (idWord) {
+    const step = this.steps.find(step => step.idWord === idWord)
+    return step ? step.token : null
   }
+
+  /**
+   * Step back
+   * @retuns { Array(Object) } - results of undone steps, for example result of unmerge action
+   */
+  undo () {
+    if (this.steps.length > 1 && this.currentStepIndex > 0) {
+      const results = this.alignToStep(this.currentStepIndex - 1)
+      return results
+    } else {
+      console.error(L10nSingleton.getMsgS('ALIGNMENT_GROUP_UNDO_ERROR'))
+    }
+  }
+
+  /**
+   * Step forward
+   */
+  redo () {
+    if (this.currentStepIndex < (this.steps.length - 1)) {
+      this.alignToStep(this.currentStepIndex + 1)
+    } else {
+      console.error(L10nSingleton.getMsgS('ALIGNMENT_GROUP_REDO_ERROR'))
+    }
+  }
+
+  /**
+   * Defines current position in the step tracker and apply/remove step actions according to the position
+   * @param {Number} stepIndex
+   * @retuns { Array(Object) } - results of undone steps, for example result of unmerge action
+   */
+  alignToStep (stepIndex) {
+    let results = [] // eslint-disable-line prefer-const
+    if (this.currentStepIndex > stepIndex) {
+      for (let i = this.currentStepIndex; i > stepIndex; i--) {
+        const result = this.removeStepAction(i)
+        if (result && result.tokensGroup) { results.push(result) }
+      }
+    } else if (this.currentStepIndex < stepIndex) {
+      for (let i = this.currentStepIndex + 1; i <= stepIndex; i++) {
+        const result = this.applyStepAction(i)
+        if (result && result.indexDeleted) { results.push(result) }
+      }
+    }
+    this.currentStepIndex = stepIndex
+    return results
+  }
+
+  /**
+   * Removes the step action,
+   * the following actions are defined - add, remove, merge
+   * @param {Number} stepIndex
+   * @retuns {Object} - results of undone steps, for now it could return only the result of unmerge action
+   */
+  removeStepAction (stepIndex) {
+    const step = this.steps[stepIndex]
+    const token = step.token
+
+    let tokenIndex
+    if (step.type === 'add' || step.type === 'remove') {
+      tokenIndex = this[token.textType].findIndex(tokenId => tokenId === token.idWord)
+    }
+
+    switch (step.type) {
+      case 'add' :
+        this[token.textType].splice(tokenIndex, 1)

```

Comment:
It is done for being more universal, for now we always undo/redo directly. But we could need to move more than one step in future, that's why I created these methods `alignToStep`, `removeStepAction`, `applyStepAction` more abstract.
Do you think that we should obligatorily define only direct way?

Reply:
not necessarily, I just felt there was a little asymmetry there with undoing a remove being a push and was trying to understand. Can you add a code comment to explain the disparity?

Reply:
Oh, I understood your thoughts.
It is so because - when we delete from the group's tokens list - the token in theory could be not only the last.
And we need to delete exactly that token that is placed on the index.
But when we need to add - we could add in any place in the list. Because the order in this list doesn't influence at all.
We have checks only - do we have the token.idWord or not here.

Will add comments

Reply:
done